index.js

```diff
@@ -0,0 +1,124 @@
+class ConcurrencyQueue {
+
+  constructor(concurrency = 0) {
+    this.concurrency = concurrency;
+    this.queue = [];
+    this.count = 0;
+    this.stopped = true;
+  }
+
+  start() {
+    if (this.stopped) {
+      this.stopped = false;
+      this.run();
+    }
+  }
+
+  async push(task) {
+    this.queue.push(task);
+    this.start();
+  }
+
+  runTask() {
+    const task = this.queue.shift();
+    this.count += 1;
+    const promise = new Promise((resolve, reject) => {
+      try {
+        resolve(task());
+      } catch (error) {
+        reject(new Error(error));
+      }
+      
+    });
+    promise.then(() => {
+      this.count -= 1; 
+      this.start();
+    })
+      .catch((err) => {
+        console.error(err);
+        this.count -= 1;
+        this.stopped = true;
+      });
+  }
+
+  run() {
+    while (this.queue.length > 0 && !this.stopped) {
+      if (this.count >= this.concurrency && this.concurrency !== 0) {
+        this.stopped = true;
```

Comment:
why do you want to stop the execution if there are enough tasks? Just do nothing.